edgedb.go

```diff
@@ -74,36 +81,98 @@ func connectOne(ctx context.Context, cfg *connConfig, conn *baseConn) error {
 		err error
 	)
 
+	conn.writer = buff.NewWriter()
+
 	for _, addr := range cfg.addrs { // nolint:gocritic
-		// todo do error values need to be checked?
 		conn.conn, err = d.DialContext(ctx, addr.network, addr.address)
 		if err != nil {
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err)
+			continue
+		}
+
+		toBeDeserialized := make(chan *soc.Data, 2)
+		r := buff.NewReader(toBeDeserialized)
+		go soc.Read(conn.conn, soc.NewMemPool(4, 256*1024), toBeDeserialized)
+
+		err = conn.setDeadline(ctx)
+		if err != nil {
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
-		err = conn.connect(ctx, cfg)
+		err = conn.connect(r, cfg)
 		if err != nil {
-			_ = conn.conn.Close()
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
+		conn.acquireReaderSignal = make(chan struct{}, 1)
+		conn.readerChan = make(chan *buff.Reader, 1)
+		conn.releaseReader(r)
 		return nil
 	}
 
 	conn.conn = nil
 	return err
 }
 
+func (c *baseConn) setDeadline(ctx context.Context) error {
+	deadline, _ := ctx.Deadline()
+	if e := c.conn.SetDeadline(deadline); e != nil {
+		return fmt.Errorf("%v", e)
+	}
+
+	return nil
+}
+
+func (c *baseConn) acquireReader(ctx context.Context) (*buff.Reader, error) {
+	c.acquireReaderSignal <- struct{}{}
+
+	select {
+	case r := <-c.readerChan:
+		if r.Err != nil {
+			return nil, r.Err
+		}
+
+		return r, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (c *baseConn) releaseReader(r *buff.Reader) {
+	go func() {
+		for r.Next(c.acquireReaderSignal) {
+			if e := c.fallThrough(r); e != nil {
+				// todo what is the right thing to do here?
+				panic(e)
```

Comment:
The right thing would be to set some internal flag "client.failed" to "true" and close all IO. Then error out to the user when they try to use any client API that the client has been closed because of an unexpected message in the protocol.

Reply:
:+1: created new issue here https://github.com/edgedb/edgedb-go/issues/39